src/exception.c

```diff
@@ -289,103 +351,85 @@ void __onCriticalException(volatile reg_block_t* regs)
 }
 
 /**
- * @brief Register a handler that will be called when the user
- *        presses the RESET button. 
- * 
- * The N64 sends an interrupt when the RESET button is pressed,
- * and then actually resets the console after about ~500ms (but less
- * on some models, see #RESET_TIME_LENGTH).
+ * @brief Register a handler that will be called when a syscall exception
  * 
- * Registering a handler can be used to perform a clean reset.
- * Technically, at the hardware level, it is important that the RCP
- * is completely idle when the reset happens, or it might freeze
- * and require a power-cycle to unfreeze. This means that any
- * I/O, audio, video activity must cease before #RESET_TIME_LENGTH
- * has elapsed.
+ * This function allows to register a handler to be invoked in response to a
+ * syscall exception, generated by the SYSCALL opcode. The opcode allows to
+ * specify a 20-bit code which, in a more traditional operating system architecture,
+ * corresponds to the "service" to be called.
  * 
- * This entry point can be used by the game code to basically
- * halts itself and stops issuing commands. Libdragon itself will
- * register handlers to halt internal modules so to provide a basic
- * good reset experience.
+ * When the registered handler returns, the execution will resume from the
+ * instruction following the syscall one.
  * 
- * Handlers can use #exception_reset_time to read how much has passed
- * since the RESET button was pressed.
+ * To allow for different usages of the code field, this function accepts
+ * a range of codes to associated with the handler. This allows a single handler
+ * to be invoked for multiple different codes, to specialize services.
  * 
- * @param cb    Callback to invoke when the reset button is pressed.
- * 
- * @note  Reset handlers are called under interrupt.
+ * @note Syscall codes in the range 0x00000 - 0x0FFFF are reserved to libdragon
+ * itself. Use a code outside that range to avoid conflicts with future versions
+ * of libdragon.
  * 
+ * @param handler  		Handler to invoke when a syscall exception is triggered
+ * @param first_code 	First syscall code to associate with this handler (begin of range)
+ * @param last_code 	Last syscall code to associate with this handler (end of range)
  */
-void register_reset_handler( void (*cb)(void) )
+void register_syscall_handler( syscall_handler_t handler, uint32_t first_code, uint32_t last_code )
 {
-	for (int i=0;i<MAX_RESET_HANDLERS;i++)
-	{		
-		if (!__prenmi_handlers[i])
+	assertf(first_code <= 0xFFFFF, "The maximum allowed syscall code is 0xFFFFF (requested: %05lx)\n", first_code);
+	assertf(last_code <= 0xFFFFF, "The maximum allowed syscall code is 0xFFFFF (requested: %05lx)\n", first_code);
+	assertf(first_code <= last_code, "Invalid range for syscall handler (first: %05lx, last: %05lx)\n", first_code, last_code);
+
+	for (int i=0;i<MAX_SYSCALL_HANDLERS;i++)
+	{
+		if (!__syscall_handlers[i].handler)
```

Comment:
Should we assert if it is registered a second time?

Reply:
done